File: SpearoGo/Views/TodayPage.swift
import SwiftUI

/// Today's weather, over two screenfuls scrolled vertically.
///
/// Everything fitted on one screen only by being clipped: the header ran under
/// the status area and the daylight row fell off the bottom of the display.
/// Each half is now sized to the viewport, so a crown turn lands squarely on a
/// complete screen rather than halfway through a row.
///
/// Every value is optional and rendered as "—" when the API does not report it.
/// Nothing on this page is substituted or estimated.
struct TodayPage: View {
    @Environment(AppState.self) private var appState

    var body: some View {
        ScrollView {
            VStack(spacing: 0) {
                nowScreen
                    .containerRelativeFrame(.vertical)

                if appState.weatherData != nil {
                    skyScreen
                        .containerRelativeFrame(.vertical)
                }
            }
        }
        .brandPage()
    }

    // MARK: - Screen 1 — what it is doing right now

    private var nowScreen: some View {
        VStack(spacing: Brand.Spacing.item) {
            Text("Today")
                .brandSectionHeader()

            if let weather = appState.weatherData {
                if let condition = weather.conditionLabel {
                    Text(condition).captionStyle()
                }

                Text(weather.airTemp.map { String(format: "%.0f°C", $0) } ?? "—")
                    .verdictStyle(color: Brand.Colors.textPrimary)
                    .accessibilityLabel(temperatureLabel(weather))

                HStack(spacing: 20) {
                    ConditionItem(icon: "thermometer.high",
                                  label: "High",
                                  value: weather.tempMax.map { String(format: "%.0f", $0) } ?? "—",
                                  unit: weather.tempMax == nil ? "" : "°")
                    ConditionItem(icon: "thermometer.low",
                                  label: "Low",
                                  value: weather.tempMin.map { String(format: "%.0f", $0) } ?? "—",
                                  unit: weather.tempMin == nil ? "" : "°")
                }
            } else if appState.isLoading {
                SkeletonBlock(width: 100, height: 28)
                SkeletonBlock(width: 130, height: 20)
                    .accessibilityLabel("Loading today's weather")
            } else {
                Text("No weather data for this spot")
                    .captionStyle()
                    .multilineTextAlignment(.center)
            }
        }
        .padding(.horizontal, Brand.Spacing.page)
    }

    // MARK: - Screen 2 — rain, cloud, and daylight

    private var skyScreen: some View {
        VStack(spacing: Brand.Spacing.section) {
            Text("Sky")
                .brandSectionHeader()

            if let weather = appState.weatherData {
                HStack(spacing: 20) {
                    ConditionItem(icon: "umbrella",
                                  label: "Rain",
                                  value: weather.precipitationChance.map(String.init) ?? "—",
                                  unit: weather.precipitationChance == nil ? "" : "%")
                    ConditionItem(icon: "cloud",
                                  label: "Cloud",
                                  value: weather.cloudCover.map(String.init) ?? "—",
                                  unit: weather.cloudCover == nil ? "" : "%")
                }
            }

            if let solunar = appState.solunarData,
               let sunrise = solunar.sunrise, let sunset = solunar.sunset {
                Text("☀️ \(timeString(sunrise))   🌙 \(timeString(sunset))")
                    .captionStyle()
                    .lineLimit(1)
                    .accessibilityLabel("Sunrise \(timeString(sunrise)), sunset \(timeString(sunset))")
            } else {
                // The Sun genuinely does not rise or set on some days at high
                // latitude. Say so rather than leaving a blank.
                Text("No sunrise or sunset today")
                    .captionStyle()
                    .multilineTextAlignment(.center)
            }
        }
        .padding(.horizontal, Brand.Spacing.page)
    }

    private func timeString(_ date: Date) -> String {
        let f = DateFormatter()
        f.timeStyle = .short
        f.dateStyle = .none
        return f.string(from: date)
    }

    private func temperatureLabel(_ weather: WeatherData) -> String {
        guard let temp = weather.airTemp else { return "Air temperature not reported" }
        return String(format: "Air temperature %.0f degrees celsius", temp)
    }
}

#Preview {
    TodayPage()
        .previewAsWatch()
        .environment(AppState.preview(verdict: .go))
}

File: wear/app/src/main/java/com/spearotracker/spearogo/ui/pages/TodayPage.kt
package com.spearotracker.spearogo.ui.pages

import androidx.compose.foundation.layout.*
import androidx.compose.foundation.pager.VerticalPager
import androidx.compose.foundation.pager.rememberPagerState
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import androidx.wear.compose.material3.Text
import com.spearotracker.spearogo.ui.AppUiState
import com.spearotracker.spearogo.ui.components.ConditionItem
import com.spearotracker.spearogo.ui.components.ConditionRow
import com.spearotracker.spearogo.ui.components.ConditionItemSkeleton
import com.spearotracker.spearogo.ui.components.ScrollingPage
import com.spearotracker.spearogo.ui.theme.Brand
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale

/**
 * Today's weather, over two vertically paged screens.
 *
 * Everything fitted on one screen only by being clipped: the header ran under
 * the system clock and the daylight row fell off the bottom of the round
 * display. A free scroll did not fix it either, because it comes to rest
 * anywhere. A snapping pager always lands square on a whole screen.
 *
 * Every value is nullable and rendered as "—" when the API does not report it.
 * Nothing on this page is substituted or estimated.
 */
@Composable
fun TodayPage(uiState: AppUiState) {
    val pagerState = rememberPagerState(pageCount = { 2 })
    val timeFormat = SimpleDateFormat("HH:mm", Locale.getDefault())
    val weather = uiState.weatherData

    VerticalPager(state = pagerState, modifier = Modifier.fillMaxSize()) { page ->
        when (page) {
            0 -> Screenful {
                Text(
                    text = "TODAY",
                    style = Brand.Typography.sectionHeader,
                    color = Brand.Colors.textSecondary,
                    modifier = Modifier.padding(bottom = Brand.Spacing.item)
                )

                when {
                    weather != null -> {
                        weather.conditionLabel?.let { condition ->
                            Text(
                                text = condition,
                                style = Brand.Typography.caption,
                                color = Brand.Colors.textSecondary,
                                textAlign = TextAlign.Center
                            )
                        }

                        Text(
                            text = weather.airTemp?.let { "%.0f°C".format(it) } ?: "—",
                            style = Brand.Typography.verdictLabel,
                            color = Brand.Colors.textPrimary,
                            modifier = Modifier.padding(vertical = Brand.Spacing.item)
                        )

                        ConditionRow {
                            ConditionItem(
                                icon = "temp",
                                label = "High",
                                value = weather.tempMax?.let { "%.0f".format(it) } ?: "—",
                                unit = if (weather.tempMax == null) "" else "°"
                            )
                            ConditionItem(
                                icon = "temp",
                                label = "Low",
                                value = weather.tempMin?.let { "%.0f".format(it) } ?: "—",
                                unit = if (weather.tempMin == null) "" else "°"
                            )
                        }
                    }

                    uiState.isLoading -> {
                        ConditionRow {
                            ConditionItemSkeleton()
                            ConditionItemSkeleton()
                        }
                    }

                    else -> {
                        Text(
                            text = "No weather data for this spot",
                            style = Brand.Typography.caption,
                            color = Brand.Colors.textSecondary,
                            textAlign = TextAlign.Center
                        )
                    }
                }
            }

            1 -> Screenful {
                Text(
                    text = "SKY",
                    style = Brand.Typography.sectionHeader,
                    color = Brand.Colors.textSecondary,
                    modifier = Modifier.padding(bottom = Brand.Spacing.item)
                )

                ConditionRow(modifier = Modifier.padding(bottom = Brand.Spacing.section)) {
                    ConditionItem(
                        icon = "rain",
                        label = "Rain",
                        value = weather?.precipitationChance?.toString() ?: "—",
                        unit = if (weather?.precipitationChance == null) "" else "%"
                    )
                    ConditionItem(
                        icon = "cloud",
                        label = "Cloud",
                        value = weather?.cloudCover?.toString() ?: "—",
                        unit = if (weather?.cloudCover == null) "" else "%"
                    )
                }

                // Daylight uses the same ConditionItem treatment as every other
                // reading, rather than a smaller caption line. The Sun genuinely
                // does not rise or set on some days at high latitude, and "—"
                // says that in the same way a missing swell reading does.
                val sunrise = uiState.solunarData?.sunrise
                val sunset = uiState.solunarData?.sunset
                ConditionRow {
                    ConditionItem(
                        icon = "sun",
                        label = "Rise",
                        value = sunrise?.let { timeFormat.format(Date(it)) } ?: "—",
                        unit = ""
                    )
                    ConditionItem(
                        icon = "moon",
                        label = "Set",
                        value = sunset?.let { timeFormat.format(Date(it)) } ?: "—",
                        unit = ""
                    )
                }
            }
        }
    }
}

/**
 * One full page of the vertical pager.
 *
 * A fixed 30.dp top inset used to clear the system clock. It did not clear the
 * round edge once the text size grew, and the page could not scroll, so a
 * large text size simply lost the daylight row off the bottom. ScrollingPage
 * carries the percentage insets and the scroll; a nested vertical scroll hands
 * the gesture to the pager once it reaches its end.
 */
@Composable
private fun Screenful(content: @Composable ColumnScope.() -> Unit) {
    ScrollingPage(content = content)
}
